### DISCUSSION TITLE
Upgrading form 3.x to 4.0.2.221

### DISCUSSION THREAD

**Discussion by CSharpJohny** [dotnet/MQTTnet#1465] [Q&A] [answered]
Hi everyone!

I'm currrently working on upgrading MQTTnet in my project from version 3.x to version 4.0.2.221. 

I stucked on one problem. Maybe you will be able to help.

In my implementation I was using `WithCommunicationTimeout(timeout)` method from `MqttClientOptionsBuilder` to specify the time that client have to communicate with server. 

In MQTTnet 4.0.2.221 `WithCommunicationTimeout` property was removed from `MqttClientOptionsBuilder` class. Instead we have `WithTimeout` property. I used it, in replace of `WithCommunicationTimeout` and hoped that the result will be the same but unfortunatelly thats not the case. 

I have unit test in which I set timeout to 1 tic to force timeout. With MQTTnet 3.x it worked correctly.
I used it like this: 
`optionBuilder.WithCommunicationTimeout(timeout) // timeout set to 1 tic`
`Task<MqttClientAuthenticateResult> result = client.ConnectAsync(optionBuilder.Build()` - then I check if the result.Status is "Faulted" and if there is an exception in result.Exception. If yes I know that the timeout occured.

In MQTTnet 4.0.2.221 implementation look a little bit different:  
`optionBuilder.WithTimeout(timeout) // timeout set to 1 tic`
`Task<MqttClientConnectResult> result = client.ConnectAsync(optionBuilder.Build())` - in the result instead of "Faulted" I get "Success" and there is no exception in result.Exception. 

Could you tell me what to change to achieve the same behaviour as in MQTTnet 3.x?

**Comment by chkr1011** [comment, marked as answer]
Please see sample "Inspect_Certificate_Validation_Errors". There is a timeout used when connecting. This how timeouts should be used in version 4.0.

**Reply by CSharpJohny** [reply]
@chkr1011  Thank you, I was able to resolve my problem with this example.
